[FIX] l10n_ch_scan_bvr: bvr_postal_in_chf with len 53

tafaRU opened the pull request:
In these following BVR strings:

0100000830759>102000304752570109843010005+ 010505863>
0100000667157>115000278565310109843010006+ 010505863>

what's is parsed as 'bvrnumber' doesn't really represent a valid bvr_adherent_num. In this case a filter domain based on acc_number is needed.

**See the message commits for further info.**

tafaRU committed:
[FIX] l10n_ch_scan_bvr

In these following BVR strings:

0100000830759>102000304752570109843010005+ 010505863>
0100000667157>115000278565310109843010006+ 010505863>

what's is parsed as 'bvrnumber' doesn't really represent a valid bvr_adherent_num.
In this case a filter domain based on acc_number is needed. Further, a postal account that refers to a bank is uniquely identified by (beneficiaire + bvrnumber).

tafaRU committed:
[FIX] l10n_ch_scan_bvr
Do not alter bank settings during the invoice creation. It is assumed they are already right and it's better to do this kind of changes in a explicit way

tafaRU commented:
@yvaucher, this PR is ready to be reviewed, so please remove label WIP.
Thanks!

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
+                        not partners_bank[0].bvr_adherent_num or
+                        partners_bank[0].bvr_adherent_num ==
+                        data['bvr_struct']['bvrnumber']):
+                    partner_bank = partners_bank[0]
```
As you check that partners_bank is on len == 1 no need to use `[0]`

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
+                        partners_bank[0].bvr_adherent_num ==
+                        data['bvr_struct']['bvrnumber']):
+                    partner_bank = partners_bank[0]
+            if len(partners_bank) > 1:
```
This could be an `elif`

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
+                    raise UserError(
+                        _('There are more than one bank corresponding '
+                          'to the current string.\n'
+                          'Please check the banks configuration.'))
```
To filter you can use `filtered` to search in the records you already fetched from database.
Plus inverse check to do only one check. Plus you can do the assignation of partner_bank only after all checks.
What would be even better is to do this in a method that returns partner_bank.

```
# remove line 370 partner_bank = partners_bank[0]
... 
    # We need to filter further by bvr_adherent_num
    partners_bank.filtered(lambda rec: rec.bvr_adherent_num == data['bvr_struct']['bvrnumber'])
    if len(partners_bank) > 1:
        raise UserError(
           _('There are more than one bank corresponding '
             'to the current string.\n'
             'Please check the banks configuration.'))
partner_bank = partners_bank
```

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
         data = {}
         data['bvr_struct'] = self._get_bvr_structurated(self.bvr_string)
+        partner_bank_model = self.env['res.partner.bank']
+        partner_bank = partner_bank_model
```
Why setting an empty record set to partner_bank ? Setting it to false isn't enough ?

tafaRU replied:
I saw it in other PRs, so I thought it could be a best practice, isn'it?
Anyway, no problem to change it.

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +345,52 @@ def validate_bvr_string(self):
             domain = [('acc_number', '=', data['bvr_struct']['beneficiaire'])]
+            partners_bank = partner_bank_model.search(domain)
+            partner_bank = (
+                partners_bank and len(partners_bank) == 1 and
```
len(empty_recordset) returns 0, you can remove `partners_bank and`

tafaRU committed:
Apply some cosmetic changes to improve the code readability

tafaRU commented:
@yvaucher, thanks for your review!
I should have addressed all your remarks, please take a look at https://github.com/OCA/l10n-switzerland/pull/222/commits/ccbdeea6daad1f967632c72914da8538f4379129

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +335,43 @@ def validate_bvr_string(self):
+                        _('There are more than one bank corresponding '
+                          'to the current string.\n'
+                          'Please check the banks configuration.'))
+                partner_bank = partners_bank
```
It seems I made few typo in my proposal.

here we want `partner_bank = partners_bank`

tafaRU replied:
> here we want partner_bank = partners_bank

And this is what I wrote, isn'it? ;-)

yvaucher replied:
You did, this variable is hardly readable or is it just me ? :-/

What I read was: `partner_bank = partner_bank` :eyes:

tafaRU replied:
:wink:

yvaucher commented on l10n_ch_scan_bvr/wizard/scan_bvr.py:
```diff
@@ -345,21 +335,43 @@ def validate_bvr_string(self):
+                    partner_bank = partners_bank
+            elif len(partners_bank) > 1:
+                # We need to filter further by bvr_adherent_num
+                partners_bank.filtered(
```
To check but I think we should assign the result to `partners_bank`

yvaucher commented:
@tafaRU Thanks for the fixes, sorry I just inserted some errors in my "code comment proposal"

tafaRU commented:
> @tafaRU Thanks for the fixes

You're welcome! I didn't understand if PR needs further fixes, could you please clarify this?

yvaucher commented:
Ok only 1 change to do is the `filtered` function, it won't alter the recordset thus you need to do:

```
recs = recs.filtered(...)
```

tafaRU commented:
@yvaucher, done, now it should be ok!

yvaucher commented:
@tafaRU Thanks :+1:
